initial draft README added (#14)
skwong25/cat-api: README.md

# Cat API #

> # A REST API using ExpressJS#

Reviewer: style nit - the extra hashtag on the right hand side isn't required

> A data repository to manage information about cats and breeds via the HTTP request-response cycle.

Reviewer: ```suggestion
A data repository to manage information about cats and pedigree breeds via the HTTP request-response cycle.
```

# Getting Started #

> Run 'npm start' on Terminal.

Reviewer: ```suggestion
Run 'npm start' on Terminal, then follow the curl commands in the How to use section
```

It'd be good to hyperlink to the how to use section here so people can get straight to testing it out.

> Run 'npm start' on Terminal.

Reviewer: Also requires `npm i` on a fresh checkout. Might also be worth mentioning that `npm test` runs the Jest/Supertest test cases

Author: This is the same as npm install, is that correct?

Reviewer: Yep. You can run `npm help` - most programs have an instruction manual which list all the possible arguments.

Access via web browser at ???

## Motivation

To create an API with full CRUD functionality to put knowledge of **Javascript** and the **ExpressJS** framework into practice.  

> It provided the opportunity to use the **mocha** framework, **jest** and **supertest** library for writing unit and integration tests. 

Reviewer: would be nice and easy to link off to these project sites

It helped me to develop my **github** workflow for better management of *branches* and *pull requests*, to enable separation of concerns for easier review and working.    

## Code Style

This project follows standard JS codestyle and *contributions should be validated locally with ESLint*.

> ## Tech/Framework Used

Reviewer: This section could probably be combined into the 'motivation' section

- [ExpressJS](https://expressjs.com/) 
- [Postman API development tool](https://www.postman.com/)
- [Jest testing framework](https://jestjs.io/), [Mocha testing framework] (https://mochajs.org/) & [Supertest Testing Library](https://www.npmjs.com/package/supertest)

> - This project was bootstrapped with ???

Reviewer: will need addressing

## Features

> The application consists of two repository classes - Cat and Breed object classes, respectively. 

Reviewer: ```suggestion
The application consists of two repository classes - `Cat` and `Breed` object classes, respectively. 
```

Would be good to highlight any code throughout the readme like this

> In addition to this, there are two 'database querying' Express Routers to handle Cat and Breed routes, respectively. 

Reviewer: ```suggestion
In addition to this, there are two Express Routers that handle Cat and Breed routes, respectively, and query a fake in-memory database. 
```

A seperate module consolidates simple functions for use across classes. All modules are then imported into the main server file.  

The API codebase is structured using repository pattern, minimising code repetition and upholding separation of concerns. 

> A consistence interface rationalises expectations for input/outputs for middleware and JS functions.  

Reviewer: ```suggestion
A consistent interface rationalises expectations for input/outputs for middleware and JS functions.  
```

> These elements support persistence ignorance, easing future implementation of a third party database. 

Reviewer: consider linking off to what persistence ignorance is

The application includes a series of unit and integration tests, strengthened by the process of Test Driven Development.
Use of **Continuous Integration** allows for automated testing. 

It increased familiarity with the processes of managing data formats, for example, parsing JSON objects, curl and url encoded data. 

> ## Code Walkthrough 

Reviewer: it'd be good to link to the file on Github's URL when giving a walkthrough. This is nice and comprehensive otherwise 👍

### Server ###

The app.js server file consists of a function, which when called instantiates an express application, then mounts helper functions and Router instances.
Helper functions consist of bodyparsing and HTTP request logger middleware.
It contains final error-handling middleware function to send error responses. 
Finally, it sets the server up at specified port 4001 listening for requests.

The app function is called in buildServer.js, passing in an id generator - shortid helper node module. 

> Seperating this step out allows a different id generator parameter to be passed in during testing. 

Reviewer: ```suggestion
Seperating this step out allows a different id generator parameter to be passed in during testing, which ensures responses are consistent and reproducible across test runs.
```

isIdValid() checks if an id is a valid shortid.

checkObject() executes a pipeline of check functions to verify that the request object is populated with valid keys and values.
An invalid id or object will generate a 400 BAD REQUEST error.

## Breed Router ##

Breed Router registers a full set of routes to enable CRUD functionality, in addition to methods to verify parameters:

isBreedIdNum() checks if an id is a number data type. 

checkBreedObj() executes a pipeline of check functions to verify that the request object is populated with valid keys and values.
Only keys matching "name" & "description" will be accepted. Only values of 'string' data type will be accepted. 
An invalid id or object will generate a 400 BAD REQUEST error.

### Repository classes ###

The cat & breed repository classes are pre-populated with properties containing cat & breeds objects, respectively. 
Each class contains getter and setter methods to retrieve and update data as a complete set. 
The subsequent methods employ the getter and setter methods, manipulating data further depending on the request.

> 

Reviewer: It's probably worth mentioning that the repository just uses an in-memory array to store cat/breed information. You can explain that in a real production application you would use a SQL database such as [sqlite](https://www.sqlite.org/index.html) to persist the data in tables.

> ```javascript

Reviewer: can be removed

```

## Cat Repository ##

> The cat repository takes an idGenerator argumen which generates a unique id for each cat object in the database upon class instantion. 

Reviewer: ```suggestion
The cat repository takes an idGenerator argument which generates a unique id for each cat object in the database upon class instantiation. 
```

                description: "buff or tan, skinny, talkative, often found in tall grasses or deep in the bush",
                breedId: 1
}
```
updateCatById() receives a cat object and a cat id. It updates only properties which differ from the existing record. Returns the updated cat object. If no matching record found, returns null. 

deleteCatById() receives a cat id. It deletes the matching cat record in the database. If no matching record found, returns null. 

addCat() receives a cat object to be added to the database. It attaches a new unique id to the object, and then updates the database via setter method. Returns the newly-added cat object.

## Validation Functions ##

Non-middleware simple JS functions exist within a separate module from the Router and Repository files:

isInvalidString() checks the data type of a passed argument. Returns a truthy in the event of a non-string. Else, returns falsey. 
isInvalidNum() checks the data type of a passed argument. Returns a truthy in the event of a non-number. Else, returns falsey. 

generateErr400() receives a string argument of a descriptive error message. Returns an 400 BAD REQUEST error.

> generateErr404() receives a string of an id which has not been found in the database. Returns an 400 NOT FOUND error.

Reviewer: It would be worth considering moving these to [inline comments](https://stackoverflow.com/a/127106), as this is very detailed and useful but might get lost to anyone reading the code instead of the README. The code could also end up getting out of date with the docs in the README.

Author: I assume you meant just those three functions? :)

Reviewer: It was more of a general comment. It doesn't matter too much in this case because the level of documentation is amazing, but might be something to watch out for in future.

The following object check functions receive an array of keys, acceptable keys and the request object as arguments:

> checkObjFormat() recchecks that the passed object is populated. 

Reviewer: ```suggestion
checkObjFormat() rechecks that the passed object is populated. 
```

```javascript
// catsRouter.js 

checkObject(req, res, next) { 

        const object = req.body;  
        const keys = Object.keys(object);
        const validKeys = ["name", "ageInYears", "favouriteToy", "description"];
        let arrayOfFunctions = [this.validate.checkObjFormat, this.validate.checkObjKeys, this.validate.checkObjValues];

        arrayOfFunctions.forEach((funct) => { 
            let message = funct(keys, object, validKeys);

            if (message[0] === "E") {
                const err = this.validate.generateErr400(message);
                return next(err);
            } else {
                console.log(message);
            };
        });
    req.object = object;  
    next()
    }
```

## Dependancies ##

Several middleware functions are employed within the app:

> morgan - logs information about HTTP requests 

```javascript
// Example of GET request logging: 
GET /cats 200 193 - 1.414 ms
```
> shortid - unique id generator

```javascript
// Example of unique id generated by shortid:    
"cats": [
        {
            "id": "UFogQKrFW",
            "name": "Catty"
        }
        ...
]
```

> > Formerly the app used body-parser middleware to parse incoming request body, available as a property of req.body. However current express version v.14.17.1 has parsing built-in: 

Reviewer: This can go as it's not useful to the reader

```javascript
// app.js
app.use(express.json()); 
```    

## Tests

To run all tests, run `npm test` in Terminal. If all tests pass, you will see the screens below:

<img src="./screenshots/testscreenshot1.png" width="450" alt="screenshot7">

<img src="./screenshots/testsScreenshot2.png" width="450" alt="screenshot7">
<img src="./screenshots/testsScreenshot3.png" width="450" alt="screenshot7">

## Unit Tests ##
    
The validation function module is tested using unit tests, run on a jest framework. Example of a typical test case:

```javascript

test('returns Error object with 400 status and given message', () => {
        expect(generateErr400('BAD REQUEST')).toHaveProperty("message", "BAD REQUEST"); 
        expect(generateErr400('BAD REQUEST')).toHaveProperty("status", 400);
    });

```

> To run the unit test suite, run `npm run unit` in Terminal. 

Reviewer: Great :) this section is nice and concise

## Integration Tests ##

Class repositories and routers are tested using integration tests. These test the request-response cycle, run on a mocha framework.
A mock id generator is passed into the constructor for enable testing assertions for requests by id: 

```javascript
const generateTestId = {
    generate() {
        return "tmk60ux2b" 
    } 
}
const appTest = buildServer(generateTestId);
```    

Example of a typical test case:

```javascript
describe('GET /cats/:id', function () {
    it('respond with json object containing detailed information on a single cat', function (done) {
        request(appTest)
            .get('/cats/tmk60ux2b')
            .set('Accept', 'application/json')
            .expect('Content-Type', /json/)
            .expect(200)
            .end((err) => {
                if (err) return done(err); 
                done(); 
            })    
    });
});
```

To run integration test suites (for both cat and breed repositories), run `npm run integration` in Terminal. 

## How to use?

> 1.  Enter 'npm start' in Terminal. The tab will now read 'node'

Reviewer: ```suggestion
1.  Enter 'npm start' in Terminal.
```
